File: backend/app/schemas/leaderboard.py
"""Schemas for weekly, season, and pick breakdown leaderboard data."""

import uuid

from app.schemas.base import CamelModel


class MemberGamePickRead(CamelModel):
    game_id: uuid.UUID
    team: str | None = None
    confidence: int | None = None


class LeaderboardMemberRead(CamelModel):
    rank: int
    member_id: uuid.UUID
    member_name: str
    total_points: int
    correct_picks: int = 0
    incorrect_picks: int = 0
    weekly_wins: int = 0
    first_place_finishes: int = 0
    second_place_finishes: int = 0
    third_place_finishes: int = 0
    payout_cents: int = 0
    points_remaining: int = 0
    last_two_game_picks: list[MemberGamePickRead] = []


class WeekLabelRead(CamelModel):
    week_number: int
    season_number: int


class GameLabelRead(CamelModel):
    game_id: uuid.UUID
    away_team: str
    home_team: str


class WeeklyLeaderboardRead(CamelModel):
    week: WeekLabelRead
    standings: list[LeaderboardMemberRead]
    last_two_games: list[GameLabelRead] = []


class SeasonStandingsRead(CamelModel):
    season: int
    standings: list[LeaderboardMemberRead]


class TeamPickCountRead(CamelModel):
    team: str
    user_count: int


class GamePickBreakdownRead(CamelModel):
    game_id: uuid.UUID
    away_team: str
    home_team: str
    away_record: str | None = None
    home_record: str | None = None
    median_confidence: float | None = None
    team_counts: list[TeamPickCountRead]


class WeeklyPickBreakdownRead(CamelModel):
    week_number: int
    games: list[GamePickBreakdownRead]


class PickBreakdownRead(CamelModel):
    season: int
    weeks: list[WeeklyPickBreakdownRead]


class GamePickDetailRead(CamelModel):
    member_name: str
    team: str
    confidence: int
    is_correct: bool | None = None


class GamePicksRead(CamelModel):
    game_id: uuid.UUID
    away_team: str
    home_team: str
    picks: list[GamePickDetailRead]

File: frontend/src/types/leaderboard.ts
export interface MemberGamePick {
  gameId: string
  team: string | null
  confidence: number | null
}

export interface LeaderboardMember {
  rank: number
  memberId: string
  memberName: string
  totalPoints: number
  correctPicks: number
  incorrectPicks: number
  weeklyWins: number
  firstPlaceFinishes: number
  secondPlaceFinishes: number
  thirdPlaceFinishes: number
  payoutCents: number
  pointsRemaining: number
  lastTwoGamePicks: MemberGamePick[]
}

export interface CompletedWeek {
  weekNumber: number
  seasonNumber: number
}

export interface GameLabel {
  gameId: string
  awayTeam: string
  homeTeam: string
}

export interface WeeklyLeaderboard {
  week: {
    weekNumber: number
    seasonNumber: number
  }
  standings: LeaderboardMember[]
  lastTwoGames: GameLabel[]
}

export interface SeasonStandings {
  season: number
  standings: LeaderboardMember[]
}

export interface TeamPickCount {
  team: string
  userCount: number
}

export interface GamePickBreakdown {
  gameId: string
  awayTeam: string
  homeTeam: string
  awayRecord: string | null
  homeRecord: string | null
  medianConfidence: number | null
  teamCounts: TeamPickCount[]
}

export interface WeeklyPickBreakdown {
  weekNumber: number
  games: GamePickBreakdown[]
}

export interface PickBreakdown {
  season: number
  weeks: WeeklyPickBreakdown[]
}

export interface GamePickDetail {
  memberName: string
  team: string
  confidence: number
  isCorrect: boolean | null
}

export interface GamePicks {
  gameId: string
  awayTeam: string
  homeTeam: string
  picks: GamePickDetail[]
}
